src/rest_framework_dso/embedding.py

```diff
@@ -0,0 +1,192 @@
+"""Logic to implement object embedding, and retrieval in a streaming-like fashion.
+
+It's written in an observer/listener style, allowing the embedded data to be determined
+during the rendering. Instead of having to load the main results in memory for analysis,
+the main objects are inspected while they are consumed by the output stream.
+"""
+from typing import Dict, Optional, Union
+
+from rest_framework import serializers
+from rest_framework.exceptions import ParseError, PermissionDenied
+
+from dso_api.dynamic_api.permissions import fetch_scopes_for_model
+from rest_framework_dso.fields import AbstractEmbeddedField, EmbeddedField
+from rest_framework_dso.serializer_helpers import ReturnGenerator, peek_iterable
+
+EmbeddedFieldDict = Dict[str, AbstractEmbeddedField]
+
+
+def get_expanded_fields(  # noqa: C901
+    parent_serializer: serializers.Serializer, fields_to_expand: Union[list, bool]
+) -> Dict[str, EmbeddedField]:
+    """Find the expanded fields in a serializer that are requested.
+    This translates the ``_expand`` query into a dict of embedded fields.
+    """
+    if not fields_to_expand:
+        return {}
+
+    allowed_names = getattr(parent_serializer.Meta, "embedded_fields", [])
+    auth_checker = getattr(parent_serializer, "get_auth_checker", lambda: None)()
+    embedded_fields = {}
+
+    # ?_expand=true should expand all names
+    auto_expand_all = fields_to_expand is True
+    if auto_expand_all:
+        fields_to_expand = allowed_names
+
+    for field_name in fields_to_expand:
+        if field_name not in allowed_names:
+            raise _expand_parse_error(allowed_names, field_name) from None
+
+        # Get the field
+        try:
+            field = getattr(parent_serializer, field_name)
+        except AttributeError:
+            raise RuntimeError(
+                f"{parent_serializer.__class__.__name__}.{field_name}"
+                f" does not refer to an embedded field."
+            ) from None
+
+        # Check access via scopes (NOTE: this is higher-level functionality)
+        scopes = fetch_scopes_for_model(field.related_model)
+        if auth_checker and not auth_checker(*scopes.table):
+            # Not allowed, silently drop for _expand=true request.
```

Comment:
Don't know what kind of logging we do in DSO-API, but things like this generally do warrant a logging statement.